### ISSUE TITLE
Xarchiver crashes at startup

### ISSUE THREAD

**Issue by nsalguero** [ib/xarchiver#20]
Hi,

On a system running Mageia Cauldron, xarchiver (latest snapshot) crashes.

After some searches, I found that, since https://github.com/ib/xarchiver/commit/4a93a6ef6805bdb583026260272008c953c06c27, launching xarchiver causes a segmentation fault because it tries to use a function named 'open' from the glibc.

The provided patch rename open to open_archive.

[xarchiver-master-revert-renaming-open_archive.txt](https://github.com/ib/xarchiver/files/976550/xarchiver-master-revert-renaming-open_archive.txt)

**Comment by ib** [comment]
I'm curious: Why would that happen? Is there a compiler warning?

**Comment by nsalguero** [comment]
There is no difference when compiling with or without the patch.

Version of:
  - GCC is 5.4.0
  - glibc is 2.22
  - gtk is 2.24.31

The attached file gives some information from GDB. 

[gdb-xarchiver-without-patch.txt](https://github.com/ib/xarchiver/files/978342/gdb-xarchiver-without-patch.txt)

**Comment by ib** [comment]
I thought so. That would have made no sense.

The debugger output shows that the crash occurs while calling xa_create_prefs_dialog() which is the first window to be created. There is a bug in that function, but I don't know whether that causes the crash.

The crash itself occurs when GTK tries to open your resource file (`/home/test/.gtkrc-2.0`). Are permissions and content okay?

Let's try to track it down with gdb. Start the debugger and type:

`break gtkrc.c:996`
Make breakpoint pending on future shared library load? y
`break gtkrc.c:1019`
Make breakpoint pending on future shared library load? y
`run`
`print rc_file->canonical_name`
`print rc_file->name`
What is displayed?
`continue`
`print rc_file->canonical_name`
What is displayed?
`next`
`next`
You should now be at `fd = g_open (rc_file->canonical_name, O_RDONLY, 0);`
`next`
Now there should be the crash when GLib calls glibc's open().

**Comment by nsalguero** [comment]
/home/test/.gtkrc-2.0 permissions are 0644 and content is generated by LXAppearance (so it seems correct).

The following file contains the GDB session: [gdb.txt](https://github.com/ib/xarchiver/files/978895/gdb.txt)

**Comment by ib** [comment]
Too much _optimized out_. Please get the latest snapshot (f0012dd), configure with `--enable-debug=full` and repeat the session.

There is an interesting part at the end of the session (prior to the crash):

>(gdb) next
>1023	      rc_file->mtime = statbuf.st_mtime;
>(gdb) next
>1025	      fd = g_open (rc_file->canonical_name, O_RDONLY, 0);
>(gdb) next
>1023	      rc_file->mtime = statbuf.st_mtime;
>(gdb) next
>1025	      fd = g_open (rc_file->canonical_name, O_RDONLY, 0);
>(gdb) next

g_open() is called twice? There is no loop in gtk_rc_context_parse_one_file().

**Comment by nsalguero** [comment]
Sorry, with the latest snapshot and "--enable-debug=full", I got the same outpout from gdb (something must force optimization).

But, yes, g_open() is called twice.

**Comment by ib** [comment]
Could you print `fd` and `rc_file->canonical_name` respectively before and after the two `1025 fd = g_open (rc_file->canonical_name, O_RDONLY, 0);`?

**Comment by ib** [comment]
> I got the same outpout from gdb (something must force optimization).

Well, of course. That's the optimization at GTK compile time.

I suppose you are using a rpm package. We could try to build it ourselves (temporarily). I can guide you if you like.

**Comment by ib** [comment]
How shall we proceed?

**Comment by nsalguero** [comment]
Sorry for the delay.  Here are my tests: I locally rebuilt the GTK package, replacing the default compilation flags ("-O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4") by "-O0 -g", installed the libraries into /usr/local and rebuilt xarchiver with "--enable-debug=full" and CFLAGS="-O0 -g".

The attached file is the result from GDB.
[gdb-gtk-debug.txt](https://github.com/ib/xarchiver/files/989318/gdb-gtk-debug.txt)

As you can see, g_open() is called only once with those compilation flags but segmentation fault occurs.

**Comment by ib** [comment]
It always seems to come down to glibc trying to open `/home/test/.gtkrc-2.0`. I'm not quite sure how to debug glibc's open, but could you try to remove `/home/test/.gtkrc-2.0` for now and see what happens?

**Comment by nsalguero** [comment]
It crashes when it tries to open ~/.Xdefaults-localhost as you can see in the GDB output.

[gdb-without-dot-gtkrc.txt](https://github.com/ib/xarchiver/files/992188/gdb-without-dot-gtkrc.txt)

**Comment by ib** [comment]
> It crashes when it tries to open ~/.Xdefaults-localhost

Which interestingly enough again is a file in /home/test (and this time the call comes from Xlib). There is seriously something wrong but I doubt it has anything to do with xarchiver. 

I never asked, but I suppose other GTK applications are running, aren't they?

Please restore your `/home/test/.gtkrc-2.0`. Save the attached test.txt in /tmp, rename it test.c and compile with `gcc -Wall -Wextra $(pkg-config glib-2.0 --cflags-only-I) $(pkg-config glib-2.0 --libs-only-l) -g -O0 test.c -o test`. If it crashes you can debug it by setting a breakpoint at test.c:12 and using the `step` command prior to the crash in order to step down to the real open() call.
If test.c runs, set the breakpoint at fcntl2.h:53 and try to step from there with the crashing xarchiver.

[Uploading test.txt…]()

**Comment by nsalguero** [comment]
Yes, other GTK applications are running.

In your test file, I corrected "/tmp/test/.gtkrc-2.0" to "/home/test/.gtkrc-2.0". The test program runs without any crash:
[test@localhost tmp]$ ./test 
/tmp/test.c: success
/home/test/.gtkrc-2.0: success

**Comment by ib** [comment]
Try setting a breakpoint at fcntl2.h:53 with the crashing xarchiver and `step` from there. (I don't know where this will lead to.)

**Comment by nsalguero** [comment]
[gdb-break-fcntl2.h_53.txt](https://github.com/ib/xarchiver/files/992532/gdb-break-fcntl2.h_53.txt)

**Comment by ib** [comment]
No new information besides:
> open (__oflag=0, __path=__path@entry=0x7fffffffc3c0 "/home/test/.Xdefaults-localhost") at /usr/include/bits/fcntl2.h:53
> **warning: Source file is more recent than executable.**

You wrote:
> Yes, other GTK applications are running.

Did you compile any of these by yourself with the same gcc and libraries you are using for xarchiver now?

Is there any xarchiver version that runs (like 0.5.4.8 or 0.5.4.7 or older)? If so, can you `git bisect` (I can guide you)?

**Comment by nsalguero** [comment]
> Did you compile any of these by yourself with the same gcc and libraries you are using for xarchiver now?

Yes, before sending packages to the Mageia build system, I often build and test them locally (I am the maintainer of LXDE for Mageia). Moreover, the build system uses the same gcc and libraries.

> Is there any xarchiver version that runs (like 0.5.4.8 or 0.5.4.7 or older)? If so, can you git bisect (I can guide you)?

Yes, version 0.5.4.8 and older worked without any problem. I found that https://github.com/ib/xarchiver/commit/4a93a6ef6805bdb583026260272008c953c06c27 is the first commit that causes the segmentation fault when running xarchiver and it seems to be also the only one because, with my patch, which "reverts" that commit, the latest git snapshot of xarchiver runs without any problem.

**Comment by ib** [comment]
Well, I don't know why I don't have any problems, but you seem right. Adding 
```
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
```
in archive.h reveals:

`main.h:40:18: error: 'open' redeclared as different kind of symbol`
`/usr/include/fcntl.h:169:12: note: previous declaration of 'open' was here`

I got confused by [your](https://github.com/ib/xarchiver/issues/20#issuecomment-299402497)
> There is no difference when compiling with or without the patch.

and thought that renaming the open_func wouldn't make any difference.

Anyway, I'm preparing a patch to test (which is tricky, because my development version is far ahead). Please stay tuned.

**Comment by ib** [comment]
Here is a patch against current master (e4b13e4). Please test.

[master.patch.txt](https://github.com/ib/xarchiver/files/993744/master.patch.txt)

**Comment by nsalguero** [comment]
Many thanks! With your patch, xarchiver runs. No more segmentation fault.

**Comment by ib** [comment]
Thanks for reporting and hanging in.

**Comment by ib** [comment]
Fixed in just now released version 0.5.4.10.
